### PR TITLE
Add projects API ordering

### PR DESCRIPTION
**Pull request by gounux**
Add an ordering filter for the `ProjectViewSet` that allows ordering by :
- `owner`'s username
- `name`
- `created_at`

See[QF-4681](https://app.clickup.com/t/2192114/QF-4681)

### CODE REVIEW

**Comment by duke-nyuki** [comment]
Task linked: [QF-4681 Add abilty to sort projects by `owner`, `name` or `created_at` ASC/DESC in the API](https://app.clickup.com/t/2192114/QF-4681)

File: docker-app/qfieldcloud/core/views/projects_views.py [hunk 1]
```diff
@@ -77,6 +77,8 @@ class ProjectViewSet(viewsets.ModelViewSet):
     lookup_url_kwarg = "projectid"
     permission_classes = [permissions.IsAuthenticated, ProjectViewSetPermissions]
     pagination_class = pagination.QfcLimitOffsetPagination()
+    filter_backends = [filters.OrderingFilter]
+    ordering_fields = ["owner__username", "name", "created_at"]
```
Comment (gounux) [inline docker-app/qfieldcloud/core/views/projects_views.py:81]: @suricactus am wondering about the `owner__username` here, does the "sort by owner" mean by username ?

Comment (suricactus) [reply docker-app/qfieldcloud/core/views/projects_views.py:81]: It does, correct!

**Comment by suricactus** [comment]
Great!

As discussed, can we make the ordering:
1)  case insensitive
2) instead of using `owner__username` parameter value, find a way to pass only `owner` and still order by username?

File: docker-app/qfieldcloud/core/views/projects_views.py [hunk 2]
```diff
@@ -46,6 +49,43 @@ def has_permission(self, request, view):
...
+            return queryset
+
+        new_ordering = []
+        for field in ordering:
+            descending = field.startswith("-")
+            clean_field = field.lstrip("-")
+
+            # owner's username is a FK
+            if field == "owner":
+                field = "owner__username"
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/views/projects_views.py:72]: Can we try to make this more generic? What if the `field` is a `dict` that has a mapping `{"owner": "owner__username"}`? Or what if it s a string that has some kind of separator `owner:owner__username`?

File: docker-app/qfieldcloud/core/drf_utils.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+        for field_name, _verbose_name in self.get_valid_fields(
+            queryset, view, context={"request": request}
+        ):
+            # standard handling of fields from the base class
+            query_field_name = self._get_query_field(base_fields, field_name)
+
+            if query_field_name:
+                valid_fields.append(query_field_name)
+                continue
```
Comment (gounux) [inline docker-app/qfieldcloud/core/drf_utils.py:85]: Hi @suricactus , added this part after our session, after having noted that sorting by `-name` or `-created_at` was not sorting as expected

File: docker-app/qfieldcloud/core/drf_utils.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    It is possible to use an `ordering_fields` expression value with attributes
+    e.g.: "my_field::alias=my_field_alias,key=value"
+    """
+
+    SEPARATOR = "::"
+    TOKENS_LIST_SEPARATOR = ","
+    TOKENS_VALUE_SEPARATOR = "="
+
+    def _get_query_field(self, fields: list[str], term: str) -> str | None:
+        """
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/drf_utils.py:19]: Generic comment for the documentation:
- please add a dot at the end of the sentences :)
- Make sure there the first line is a concise summary of what the function is doing, add a new empty line and continue with a more detailed explanation.
- Follow with arguments and return value documentation too.

See some docs examples from here: https://github.com/opengisch/QFieldCloud/blob/master/docker-app/qfieldcloud/core/utils2/storage.py

Comment (gounux) [reply docker-app/qfieldcloud/core/drf_utils.py:19]: Sure, let's follow the convention

File: docker-app/qfieldcloud/core/drf_utils.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+        Custom attributes expression has form : "my_field::alias=my_field_alias,key=value".
+        In the above example, `alias` is the URL GET param value,
+        but `my_field` is the real model field.
+
+        Args:
+            queryset (QuerySet): Django's Project queryset
+            fields (list[str]): HTTP query's ordering fields
+            view (ModelViewSet): DRF's Project view
+            request (HttpRequest): incoming HTTP request
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/drf_utils.py:90]:
```suggestion
            queryset (QuerySet): Django's ORM queryset
            fields (list[str]): ordering fields passed to the HTTP querystring
            view (ModelViewSet): DRF view instance
            request (HttpRequest): DRF request instance
```

Note that the Project queryset and the Project view are a special usecase, the Filter should be generic.

File: docker-app/qfieldcloud/core/drf_utils.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        Args:
+            fields (list[str]): list of fields to search
+            term (str): term to search in the list
+        Returns:
+            str | None: the matching field, if present in the list
+        """
+        for field in fields:
+            compare_value = field
+
+            if field.startswith("-"):
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/drf_utils.py:33]:
```suggestion
            # the "-" is used when the fields are sorted descending
            if field.startswith("-"):
```

**Review by suricactus — APPROVED** [review]
Ready too merge after the last few small comments. Please squash and force push to a single commit.

**Review by suricactus — APPROVED** [review]

**Pull request merged**